node/src/components/linear_chain_sync.rs

```diff
@@ -26,114 +50,246 @@ pub trait ReactorEventT<I>:
+        };
+        remove_peer(&mut self.peers);
+        remove_peer(&mut self.peers_to_try);
     }
 
-    fn new_block(&mut self, block: Block) {
-        self.linear_chain.push(block);
-        self.linear_chain_length += 1;
+    /// Add new block to linear chain.
+    fn add_block(&mut self, block: Block) {
+        match &mut self.state {
+            State::None | State::Done => {}
+            State::SyncingTrustedHash { linear_chain, .. } => linear_chain.push(block),
+            State::SyncingDescendants { linear_chain, .. } => linear_chain.push_back(block),
+        };
     }
 
     /// Returns `true` if we have finished syncing linear chain.
     pub fn is_synced(&self) -> bool {
-        self.is_synced
+        match self.state {
+            State::None | State::Done => true,
+            _ => false,
+        }
     }
 
-    fn fetch_next_block_deploys<R, REv>(
+    /// Handle an event indicating that a linear chain block has been handled by consensus
+    /// component.
+    /// Returns effects that are created as a response to that event.
+    fn block_handled<R, REv>(
         &mut self,
-        effect_builder: EffectBuilder<REv>,
         rng: &mut R,
+        effect_builder: EffectBuilder<REv>,
+        block_height: u64,
     ) -> Effects<Event<I>>
     where
         I: Send + Copy + 'static,
         R: Rng + CryptoRng + ?Sized,
         REv: ReactorEventT<I>,
     {
-        let peer = self.random_peer_unsafe(rng);
-        match self.linear_chain.pop() {
+        let curr_state = mem::replace(&mut self.state, State::None);
+        let (new_state, mut effects) = match curr_state {
+            State::None | State::Done => panic!("Block handled when in {:?} state.", &curr_state),
+            State::SyncingTrustedHash {
+                highest_block_seen,
+                trusted_hash,
+                ..
+            } => {
+                if block_height == highest_block_seen {
+                    info!(%block_height, "Finished synchronizing linear chain up until trusted hash.");
+                    self.reset_peers(rng);
+                    let peer = self.random_peer_unsafe();
+                    // Kick off syncing trusted hash descendants.
+                    let effects = fetch_block_at_height(effect_builder, peer, block_height + 1);
+                    (State::sync_descendants(trusted_hash), effects)
+                } else {
+                    (curr_state, Effects::new())
+                }
+            }
+            State::SyncingDescendants {
+                highest_block_seen, ..
+            } => {
+                if block_height == highest_block_seen {
+                    info!(%block_height, "Finished synchronizing descendants of trusted hash.");
+                    (State::Done, Effects::new())
+                } else {
+                    (curr_state, Effects::new())
+                }
+            }
+        };
+        self.state = new_state;
+        if !self.state.is_done() {
```

Comment:
`panic!` in `fetch_next_block_deploys` indicates that there's an issue with timing and order of execution/handling events.

If we're not `Done` yet, it means that we're still in either `SyncingTrustedHash` or `SyncingDescendants` state. If there are more blocks we haven't yet executed we will ask for their deploys but if we've consumed whole `linear_chain` then it will return empty vector of effects and we will wait until all blocks are handled.